File: src/ct/src/item.py
from ct.src.engine import SDL_RenderCopy
from ct.src.image import *
from ct.src.utils import SDL_Point, SDL_Rect, create_point, tile_point_to_world_point, create_rect, get_move_grid_size_dims, set_scaled_rect, set_scaled_rect_with_camera
from ct.src.image import Image
from ct.src.ct_array import Array
from ct.src.base_types import Uint32, addr, Uint64, cast, last_enum_int
from ct.src.sprite import Sprite
from ct.src.constants import BuildingName, ImageName, ItemName, ItemType
# weak imports
from ct.src.game import Game


class Item:
    def __init__(self):
        self.item_name: ItemName = None
        self.item_type: ItemType = None
        self.sprite: Sprite = None
        self.cost: Uint64 = None

    def update(self, game: Game):
        self.sprite.update(game)

    def draw(self, game: Game):
        self.sprite.draw(game)

    def set_to(self, game: Game, item_name: ItemName):
        # clear previous data
        self.sprite.srcs.clear()
        # set new data
        self.item_name = item_name
        self.sprite.spawn_time = game.engine.current_time
        self.sprite.render_dst.is_camera_rendered = True
        self.item_type = ItemType.Misc

        if item_name == ItemName.NA:
            print("Item.set_to - attempted to call with ItemName.NA.")
            exit(1)
        elif item_name == ItemName.Jelly:
            self.sprite.image = game.engine.get_image(ImageName.Items)
            self.sprite.anim_speed = 100
            src0 = create_rect(20, 0, 20, 20)
            self.sprite.srcs.push(src0)
            self.cost = 17
        elif item_name == ItemName.FurPelt:
            self.sprite.image = game.engine.get_image(ImageName.Items)
            self.sprite.anim_speed = 100
            src0 = create_rect(40, 0, 20, 20)
            self.sprite.srcs.push(src0)
            self.cost = 32
        elif item_name == ItemName.Branch:
            self.sprite.image = game.engine.get_image(ImageName.Items)
            self.sprite.anim_speed = 100
            src0 = create_rect(60, 0, 20, 20)
            self.sprite.srcs.push(src0)
            self.cost = 44
        elif item_name == ItemName.Mushroom:
            self.sprite.image = game.engine.get_image(ImageName.Items)
            self.sprite.anim_speed = 100
            src0 = create_rect(80, 0, 20, 20)
            self.sprite.srcs.push(src0)
            self.cost = 38
        elif item_name == ItemName.BatWing:
            self.sprite.image = game.engine.get_image(ImageName.Items)
            self.sprite.anim_speed = 100
            src0 = create_rect(100, 0, 20, 20)
            self.sprite.srcs.push(src0)
            self.cost = 27
        elif item_name == ItemName.Apple:
            self.item_type = ItemType.Useable
            self.sprite.image = game.engine.get_image(ImageName.Items)
            self.sprite.anim_speed = 100
            src0 = create_rect(120, 0, 20, 20)
            self.sprite.srcs.push(src0)
            self.cost = 50
        else:
            item_name_as_int = cast(item_name, "int")
            print("Item.set_to. Item name not handled. ItemName:", item_name_as_int)
            exit(1)

File: src/ct/game/item.c
#include <SDL.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>
#include "ct_stacktrace.h"
#include "item.h"
#include "engine.h"
#include "image.h"
#include "utils.h"
#include "image.h"
#include "ct_array.h"
#include "sprite.h"
#include "constants.h"
#include "game.h"

void void_mfunc_Item_set_to_ItemPtr_GamePtr_ItemName(CT_StackTrace* ct_stacktrace, int func_idx_of_call, int line_number_of_call, Item* self, Game* game, ItemName item_name) {
    int item_name_as_int = 0;
    SDL_Rect src0 = {0};

    ct_stacktrace_push(ct_stacktrace, func_idx_of_call, line_number_of_call);
    void_clear_Array_SDL_Rect_10Ptr(ct_stacktrace, 10, 25, &self->sprite.srcs);
    self->item_name = item_name;
    self->sprite.spawn_time = game->engine.current_time;
    self->sprite.render_dst.is_camera_rendered = true;
    self->item_type = ItemType_Misc;
    if (item_name == ItemName_NA) {
        printf("%s\n", "Item.set_to - attempted to call with ItemName.NA.");
        ct_stacktrace_push(ct_stacktrace, 10, 34);
        ct_stacktrace_print(ct_stacktrace);
        exit(1);
}
    else if (item_name == ItemName_Jelly) {
        self->sprite.image = Image_mfunc_Engine_get_image_EnginePtr_ImageName(ct_stacktrace, 10, 36, &game->engine, ImageName_Items);
        self->sprite.anim_speed = 100;
        src0 = SDL_Rect_create_rect_int_int_int_int(ct_stacktrace, 10, 38, 20, 0, 20, 20);
        void_push_Array_SDL_Rect_10Ptr_SDL_RectPtr(ct_stacktrace, 10, 39, &self->sprite.srcs, &src0);
        self->cost = 17;
}
    else if (item_name == ItemName_FurPelt) {
        self->sprite.image = Image_mfunc_Engine_get_image_EnginePtr_ImageName(ct_stacktrace, 10, 42, &game->engine, ImageName_Items);
        self->sprite.anim_speed = 100;
        src0 = SDL_Rect_create_rect_int_int_int_int(ct_stacktrace, 10, 44, 40, 0, 20, 20);
        void_push_Array_SDL_Rect_10Ptr_SDL_RectPtr(ct_stacktrace, 10, 45, &self->sprite.srcs, &src0);
        self->cost = 32;
}
    else if (item_name == ItemName_Branch) {
        self->sprite.image = Image_mfunc_Engine_get_image_EnginePtr_ImageName(ct_stacktrace, 10, 48, &game->engine, ImageName_Items);
        self->sprite.anim_speed = 100;
        src0 = SDL_Rect_create_rect_int_int_int_int(ct_stacktrace, 10, 50, 60, 0, 20, 20);
        void_push_Array_SDL_Rect_10Ptr_SDL_RectPtr(ct_stacktrace, 10, 51, &self->sprite.srcs, &src0);
        self->cost = 44;
}
    else if (item_name == ItemName_Mushroom) {
        self->sprite.image = Image_mfunc_Engine_get_image_EnginePtr_ImageName(ct_stacktrace, 10, 54, &game->engine, ImageName_Items);
        self->sprite.anim_speed = 100;
        src0 = SDL_Rect_create_rect_int_int_int_int(ct_stacktrace, 10, 56, 80, 0, 20, 20);
        void_push_Array_SDL_Rect_10Ptr_SDL_RectPtr(ct_stacktrace, 10, 57, &self->sprite.srcs, &src0);
        self->cost = 38;
}
    else if (item_name == ItemName_BatWing) {
        self->sprite.image = Image_mfunc_Engine_get_image_EnginePtr_ImageName(ct_stacktrace, 10, 60, &game->engine, ImageName_Items);
        self->sprite.anim_speed = 100;
        src0 = SDL_Rect_create_rect_int_int_int_int(ct_stacktrace, 10, 62, 100, 0, 20, 20);
        void_push_Array_SDL_Rect_10Ptr_SDL_RectPtr(ct_stacktrace, 10, 63, &self->sprite.srcs, &src0);
        self->cost = 27;
}
    else if (item_name == ItemName_Apple) {
        self->item_type = ItemType_Useable;
        self->sprite.image = Image_mfunc_Engine_get_image_EnginePtr_ImageName(ct_stacktrace, 10, 67, &game->engine, ImageName_Items);
        self->sprite.anim_speed = 100;
        src0 = SDL_Rect_create_rect_int_int_int_int(ct_stacktrace, 10, 69, 120, 0, 20, 20);
        void_push_Array_SDL_Rect_10Ptr_SDL_RectPtr(ct_stacktrace, 10, 70, &self->sprite.srcs, &src0);
        self->cost = 50;
}
    else {
        item_name_as_int = (int)(item_name);
        printf("%s %d\n", "Item.set_to. Item name not handled. ItemName:", item_name_as_int);
        ct_stacktrace_push(ct_stacktrace, 10, 75);
        ct_stacktrace_print(ct_stacktrace);
        exit(1);
    }
    ct_stacktrace_pop(ct_stacktrace);
}

